### ISSUE TITLE
[SPIRV] HLSL lacks syntax for fully specifying formats for RWTexture2D in SPIR-V

### ISSUE THREAD

**Issue by jekstrand** [microsoft/DirectXShaderCompiler#2498]
The HLSL `RWTexture2D` syntax only allows for fairly generic format specifications where you can specify the shader data type (`float4`, `int4`, etc) and, for floating point, whether or not `unorm` is expected.  SPIR-V, on the other hand has format specifiers which provide the full API-level image format such as **Rgb10A2**.  In order to be able to properly use HLSL with Vulkan on drivers that don't support Vulkan's `shaderStorageImageReadWithoutFormat` feature, we need the fully specified format and not just `float4 unorm`.  We also need synatax (and plumbing through DXC) for SPIR-V's **NonReadable** and **NonWriteable** decorations.

In https://github.com/KhronosGroup/glslang/issues/1912, @johnkslang proposed the following syntax:

    [[spv::format_rgba8]]  // similarly for all ~39 formats; overrides the glslang deduction
    [[spv::nonreadable]]
    [[spv::nonwritable]]

I don't know which issue it's best discussed on but this should ideally be implemented in both HLSL -> SPIR-V paths.

**Comment by johnkslang** [comment]
Here is the complete list:

```
format_rgba32f
format_rgba16f
format_r32f
format_rgba8
format_rgba8snorm
format_rg32f
format_rg16f
format_r11fg11fb10f
format_r16f
format_rgba16
format_rgb10a2
format_rg16
format_rg8
format_r16
format_r8
format_rgba16snorm
format_rg16snorm
format_rg8snorm
format_r16snorm
format_r8snorm
format_rgba32i
format_rgba16i
format_rgba8i
format_r32i
format_rg32i
format_rg16i
format_rg8i
format_r16i
format_r8i
format_rgba32ui
format_rgba16ui
format_rgba8ui
format_r32ui
format_rgb10a2ui
format_rg32ui
format_rg16ui
format_rg8ui
format_r16ui
format_r8ui

nonwritable
nonreadable
```

I'm totally open to what name space to use.  Vulkan (`vk`) seems too narrow, so I'm suggesting SPIR-V (`spv`), but maybe there is an even broader one that also makes sense.

**Comment by akioCL** [comment]
Hi:

Anybody knows if there's an update on this feature? I'm having issues reading from a RWTexture2D with format rgba16f on some Android devices (Pixel 3).

Thanks!

**Comment by CaffeinePwrdAl** [comment]
Speaking on behalf of Imagination Technologies here, we're also interested in this change being made to DXC. There's some additional subtleties as to this on top of just whether shaderStorageImageWriteWithoutFormat are being relied upon by DXC:

1. the SPIR-V should be decorated with **StorageImageReadWithoutFormat/StorageImageWriteWithoutFormat** when this is expected, otherwise the format in the shader must still match the resource being stored to.

2. If the **StorageImageReadWithoutFormat/StorageImageWriteWithoutFormat** declarations are present, DXC **can** then use the SPIR-V format 'Unknown' - which indicates to the implementation to defer to the format of the image resource (Unknown is compatible with everything). I'm not sure if the spec is clear enough on whether Unknown is **required** in this case, but it can only be used when the WithoutFormat decorations are present, so obviously there was some intent there.

Otherwise I believe that in all cases where a format other than 'Unknown' is specified on the OpTypeImage (regardless of features and decorations present) an implementation **can** use the format in the shader, and thus it must match the resource.

We have a series of GPU's that performs image writes based only on the format declared in the shader (which should be safe in Vulkan). We **_can_** support both read and write WithoutFormat, but it comes at some expense, so currently that workaround is **only** applied when the format is specified as 'Unknown'.

We've seen a number of benchmarks that are using DXC hit this issue - in both cases it is an f16 image, but the DXC compiler produces SPIR-V that ignores the format declaration and just always specifies an F32 format - meaning it is not something the vendors of these benchmarks are easily able to fix.

Example code-snippet with the format declaration:

http://shader-playground.timjones.io/10f71034ada001f94d6956f45c43b9a5

**Comment by ehsannas** [comment]
Thanks for your information, @CaffeinePwrdAl . I'll try to reprioritize.

**Comment by seppala2** [comment]
This is also quite important for us and I actually tested an implementation for syntax like [[vk::image_format("Rgba16f")]]. So basically it is an attribute with EnumArgument, with the format identifiers from SPIR-V spec. The namespace could be easily changed to spv, but I'm not sure how the non-string syntax "spv::format_<x>" would be added to DXC. @ehsannas If this is of any help, I can share the code.

I would also like to add that it would be really nice if we didn't have to specify the format in function arguments, local variables etc. and only in the type of a global resource variable. To achieve this I removed a check inside SPIRV-Tools ops pass::GenerateCopy for image types, allowing different image format for the arguments of OpStore, and it seems like the legalization pass (?) then fixes the type mismatches (some format -> Unknown) by inlining everything.

Here's an example of how it works with my local changes:

Input:

```
[[vk::image_format("Rgba16f")]]
RWBuffer<float4> Buf;
[[vk::image_format("Rgba8")]]
RWBuffer<float4> Buf2;

struct S
{
    RWBuffer<float4> b;
};

float4 getVal(
    RWBuffer<float4> b)
{
    return b[0];
}
float4 getValStruct(S s)
{
    return s.b[1];
}

[numthreads(1, 1, 1)] void main()
{
    float4 test = getVal(Buf);
    test += getVal(Buf2);

    S s;
    s.b = Buf;
    test += getValStruct(s);

    S s2;
    s2.b = Buf2;
    test += getValStruct(s2);

    RWBuffer<float4> var = Buf;
    RWBuffer<float4> var2 = Buf2;
    test += var[2];
    test += var2[2];

    Buf[10] = test + 1;
}
```

Output:
```
; SPIR-V
; Version: 1.0
; Generator: Google spiregg; 0
; Bound: 39
; Schema: 0
               OpCapability Shader
               OpCapability SampledBuffer
               OpCapability ImageBuffer
               OpMemoryModel Logical GLSL450
               OpEntryPoint GLCompute %main "main"
               OpExecutionMode %main LocalSize 1 1 1
               OpSource HLSL 600
               OpName %type_buffer_image "type.buffer.image"
               OpName %Buf "Buf"
               OpName %type_buffer_image_0 "type.buffer.image"
               OpName %Buf2 "Buf2"
               OpName %main "main"
               OpDecorate %Buf DescriptorSet 0
               OpDecorate %Buf Binding 1
               OpDecorate %Buf2 DescriptorSet 0
               OpDecorate %Buf2 Binding 0
       %uint = OpTypeInt 32 0
     %uint_2 = OpConstant %uint 2
      %float = OpTypeFloat 32
    %float_1 = OpConstant %float 1
    %v4float = OpTypeVector %float 4
         %11 = OpConstantComposite %v4float %float_1 %float_1 %float_1 %float_1
    %uint_10 = OpConstant %uint 10
     %uint_0 = OpConstant %uint 0
     %uint_1 = OpConstant %uint 1
%type_buffer_image = OpTypeImage %float Buffer 2 0 0 2 Rgba16f
%_ptr_UniformConstant_type_buffer_image = OpTypePointer UniformConstant %type_buffer_image
%type_buffer_image_0 = OpTypeImage %float Buffer 2 0 0 2 Rgba8
%_ptr_UniformConstant_type_buffer_image_0 = OpTypePointer UniformConstant %type_buffer_image_0
       %void = OpTypeVoid
         %18 = OpTypeFunction %void
        %Buf = OpVariable %_ptr_UniformConstant_type_buffer_image UniformConstant
       %Buf2 = OpVariable %_ptr_UniformConstant_type_buffer_image_0 UniformConstant
       %main = OpFunction %void None %18
         %19 = OpLabel
         %20 = OpLoad %type_buffer_image %Buf
         %21 = OpImageRead %v4float %20 %uint_0 None
         %22 = OpLoad %type_buffer_image_0 %Buf2
         %23 = OpImageRead %v4float %22 %uint_0 None
         %24 = OpFAdd %v4float %21 %23
         %25 = OpLoad %type_buffer_image %Buf
         %26 = OpImageRead %v4float %25 %uint_1 None
         %27 = OpFAdd %v4float %24 %26
         %28 = OpLoad %type_buffer_image_0 %Buf2
         %29 = OpImageRead %v4float %28 %uint_1 None
         %30 = OpFAdd %v4float %27 %29
         %31 = OpLoad %type_buffer_image %Buf
         %32 = OpLoad %type_buffer_image_0 %Buf2
         %33 = OpImageRead %v4float %31 %uint_2 None
         %34 = OpFAdd %v4float %30 %33
         %35 = OpImageRead %v4float %32 %uint_2 None
         %36 = OpFAdd %v4float %34 %35
         %37 = OpFAdd %v4float %36 %11
         %38 = OpLoad %type_buffer_image %Buf
               OpImageWrite %38 %uint_10 %37 None
               OpReturn
               OpFunctionEnd

```

**Comment by ehsannas** [comment]
Thanks @seppala2 . I actually think `vk::image_format("Rgba16f")` is a better format than `spv::format_rgba16f`. Would you please share your code by posting a pull-request?

**Comment by seppala2** [comment]
@ehsannas sure, what is the preferred workflow since this requires a small change in SPIRV-Tools as well? Make a pull request to SPIRV-Tools, and once it gets merged, make the pull request here (with updated spirv-tools), I assume?

**Comment by ehsannas** [comment]
Yes, that's the correct workflow.

But in this particular case, to determine whether the SPIRV-Tools change is reasonable, I'd need to know the code changes made in DXC as well. So for now I just want to inspect the code and do some testing on my own machine (I don't want to wait for the SPIRV-Tools PR to land). So would you make 2 PRs (one against SPIRV-Tools repo, and one for "non-spirv-tools" changes against DXC repo), so that I can see both of the PRs at the same time and manually test a few things locally?

If we find that's the right way to go, we'll land the SPIRV-Tools PR and we'll import it into DXC and we'll rebase the second PR and make things work :)

P.S. Hope my comment here wasn't confusing.

**Comment by seppala2** [comment]
@ehsannas I've created the pull requests now:
https://github.com/microsoft/DirectXShaderCompiler/pull/3395
https://github.com/KhronosGroup/SPIRV-Tools/pull/4126

Specifically the change in SPIRV-Tools makes me think that maybe there is some better way to do it, but that seems to work quite well, as seen in the example above. You can test that without that change in spirv-tools opt, everything works fine until you try to assign the texture/buffer to a variable or pass it to a funtion. Then it will fail into the assertion about the type mismatch in Pass::GenerateCopy during FixStorageClass since the image formats don't match.

**Comment by ehsannas** [comment]
Thanks for the detailed info @seppala2 and for making the PRs. I'll take a closer look at both PRs.

**Comment by jaebaek** [comment]
This issue was fixed by #3395. Let me close this bug. Please reopen it if there are some remaining issues.
